### PR TITLE
Fix dissection of TLS communication with NULL Cipher.

### LINKED ISSUE #1976: Dissecting TLS with NULL cipher
#### Brief description
Scapy incorrectly dissects TLS packets with NULL cipher - message dissecting is too greedy which results in consuming MAC (and possibly padding)

#### Environment

-   Scapy version: ` e44bd373f0839ed67c1c91cbfc94cde58aee8722`
-   Python version: `2.7`
-   Operating System: `Windows`

#### How to reproduce
Execute following code with attached [pcap](https://github.com/secdev/scapy/files/3047186/handshake.zip):

```python
from scapy.all import *
from scapy.layers.tls.record import TLS

bind_layers(TCP, TLS)

CLIENT_IP = '10.192.71.100'
SERVER_IP = '10.192.71.150'
from_pcap = rdpcap("handshake.pcapng")

session = None
tls_only = []
for pkt in from_pcap:
    if pkt.haslayer('TLS'):
        if not session:
            session = pkt[TLS].tls_session
        else:
            connection_end = session.rcs.connection_end
            # depending on who is the target switch TLS session state
            if ((pkt[IP].dst == SERVER_IP and connection_end == 'client') or
                (pkt[IP].dst == CLIENT_IP and connection_end == 'server')):
                session.mirror()
        # dissect again TLS with session context
        tls_only.append(TLS(raw(pkt[TLS]), tls_session=session))
tls_only[5].show()
```
The pcap contains TLS handshake:
image

I marked a TLS record which is a subject of the analysis.

#### Actual result

```
###[ TLS ]### 
        type      = handshake
        version   = TLS 1.2
        len       = 36
        iv        = ''
        \msg       \
         |###[ TLS Handshake - Finished ]### 
         |  msgtype   = finished
         |  msglen    = 12
         |  vdata     = 'YX\xacX\xb81\x1fX\x8f\xbe\x1dJ'
         |###[ TLS Handshake - Client Key Exchange ]### 
         |  msgtype   = client_key_exchange
         |  msglen    = 13552178
         |  \exchkeys  \
         |   |###[ Raw ]### 
         |   |  load      = '\xb4\xc3m\xf1\x16c\xdb\xfc\x08\x16\x1d\x82\x83U\x8c\xe1'
        mac       = ''
        pad       = ''
        padlen    = None
```

#### Expected result

```
###[ TLS ]### 
        type      = handshake
        version   = TLS 1.2
        len       = 36    [deciphered_len= 16]
        iv        = ''
        \msg       \
         |###[ TLS Handshake - Finished ]### 
         |  msgtype   = finished
         |  msglen    = 12
         |  vdata     = 'YX\xacX\xb81\x1fX\x8f\xbe\x1dJ'
        mac       = '\x10\xce\xca2\xb4\xc3m\xf1\x16c\xdb\xfc\x08\x16\x1d\x82\x83U\x8c\xe1'
        pad       = ''
        padlen    = None
```

#### Related resources
[handshake.zip](https://github.com/secdev/scapy/files/3047186/handshake.zip)

#### Fix proposal
I'v got expected result by removing `and not isinstance(self.tls_session.rcs.cipher, Cipher_NULL)` from https://github.com/secdev/scapy/blob/e44bd373f0839ed67c1c91cbfc94cde58aee8722/scapy/layers/tls/record.py#L488
The fix seems to be too easy so I doubt in its correctness. Please provide guidance.

### PR DESCRIPTION
**Pull request by wasilukm**
Due to invalid handling of TLS record length, fields (MSG and MAC) were
incorrectly interpreted in packets encrypted with NULL Cipher.

fixes #1976

### CODE REVIEW

**Comment by gpotter2** [comment]
It would be great if you could add a unit test for that. For instance a packet to dissect. See the .uts files in test/.

The format is custom (historic reasons), but quite self explicit :smile:

**Comment by wasilukm** [comment]
Yeah I'm woring on UT.
BTW. I wonder why coverage report is so bad.

**Comment by gpotter2** [comment]
The Travis tests didn't run for some reason. It probably will be fixed the next time you push

File: scapy/layers/tls/cert.py [hunk 1]
```diff
@@ -534,7 +534,7 @@ def import_from_asn1pkt(self, privkey):
         self.pubkey = self.key.public_key()
 
     @crypto_validator
-    def verify(self, msg, sig, h="sha256", **kwargs):
+    def verify(self, msg, sig, t="pkcs", h="sha256", mgf=None, L=None):
```
Comment (gpotter2) [inline scapy/layers/tls/cert.py]: Please undo these changes. Even if they trigger a warning in Codacy (which we'll ignore), they add unused zrguments

Comment (wasilukm) [reply scapy/layers/tls/cert.py]: This causes an exception for my example handshake with TLS_ECDHE_ECDSA_WITH_NULL_SHA, see https://ci.appveyor.com/project/secdev/scapy/builds/23842720/job/m09g9p06yrna1l4b
Please let me know if this could be fixed in different way.

Comment (wasilukm) [reply scapy/layers/tls/cert.py]: I will try to fix this without changing methods signature.

Comment (gpotter2) [reply scapy/layers/tls/cert.py]: I guess updating https://github.com/secdev/scapy/blob/cda583a17953964a08e21509103a193a0d759c48/scapy/layers/tls/cert.py#L304 and https://github.com/secdev/scapy/blob/cda583a17953964a08e21509103a193a0d759c48/scapy/layers/tls/cert.py#L513 to use keywords arguments will do

File: scapy/layers/tls/cert.py [hunk 2]
```diff
@@ -671,7 +671,7 @@ def encrypt(self, msg, t="pkcs", h="sha256", mgf=None, L=None):
         return self.pubKey.encrypt(msg, t, h, mgf, L)
 
     def verify(self, msg, sig, t="pkcs", h="sha256", mgf=None, L=None):
-        return self.pubKey.verify(msg, sig, t, h, mgf, L)
+        return self.pubKey.verify(msg, sig, t=t, h=h, mgf=mgf, L=L)
```
Comment (wasilukm) [inline scapy/layers/tls/cert.py]: Changing this seems to be sufficient, at least in my case :)

Comment (gpotter2) [reply scapy/layers/tls/cert.py]: Could you please change both anyways ? It's better that way

Comment (wasilukm) [reply scapy/layers/tls/cert.py]: Sure, shall I do the same for `encrypt` methods. They seem to have the same pattern.

Comment (wasilukm) [reply scapy/layers/tls/cert.py]: I've updated calls of `verify`, `encrypt` and `sign`.

**Comment by wasilukm** [comment]
@gpotter2 is anything else which needs to be addressed? Could you please consider approving the change?

**Comment by gpotter2** [comment]
Looks okay to me, though I'm not familiar enough with TLS to confirm it won't break anything.
Other maintainers will merge it :-)

**Comment by gpotter2** [comment]
I had another look at this. The fix you've done seems wrong.

The code clearly says the `deciphered_len` should only be defined when something was actually deciphered (not the case with a null encoding), therefore the issue is more likely in the place that uses `deciphered_len`

**Comment by wasilukm** [comment]
@gpotter2 thanks for the feedback.
I would assume that null cipher automatically becomes deciphered but I may be wrong as I'm not TLS expert.
Do you have any  clue where to look?

**Comment by guedou** [comment]
I am suite unsure that this is a correst fix.
